AlphaWallet/TokenScriptClient/Coordinators/EventSourceCoordinator.swift

```diff
@@ -0,0 +1,149 @@
+        //TODO when TokenScript schema allows it, support more than 1 filter
+        let filterTextEquivalent = filterParam.compactMap({ $0?.textEquivalent }).first
+        let filterText = filterTextEquivalent ?? "\(eventOrigin.eventFilter.name)=\(eventOrigin.eventFilter.value)"
+        return EventInstance(contract: eventOrigin.contract, tokenContract: token.contractAddress, server: server, eventName: eventOrigin.eventName, blockNumber: Int(blockNumber), logIndex: Int(logIndex), filter: filterText, json: json)
+    }
+
+    private func convertToJsonCompatible(dictionary: [String: Any]) -> [String: Any] {
+        Dictionary(uniqueKeysWithValues: dictionary.compactMap { key, value -> (String, Any)? in
+            switch value {
+            case let address as EthereumAddress:
+                return (key, address.address)
+            case let data as Data:
+                return (key, data.hexEncoded)
+            case let string as String:
+                return (key, string)
+            case let bigUInt as BigUInt:
+                return (key, Int(bigUInt))
+            default:
+                //We only accept known types, otherwise serializing to JSON will crash
+                return nil
+            }
+        })
+    }
+
+    private func formFilterFrom(fromParameter parameter: EventParameter, tokenId: TokenId, filterName: String, filterValue: String) -> (filter: [EventFilterable], textEquivalent: String)? {
+        guard parameter.name == filterName else { return nil }
+        guard let parameterType = SolidityType(rawValue: parameter.type) else { return nil }
+        let optionalFilter: (filter: AssetAttributeValueUsableAsFunctionArguments, textEquivalent: String)?
+        if let implicitAttribute = EventSourceCoordinator.convertToImplicitAttribute(string: filterValue) {
+            switch implicitAttribute {
+            case .tokenId:
+                optionalFilter = AssetAttributeValueUsableAsFunctionArguments(assetAttribute: .uint(tokenId)).flatMap { (filter: $0, textEquivalent: "\(filterName)=\(tokenId)") }
+            case .ownerAddress:
+                optionalFilter = AssetAttributeValueUsableAsFunctionArguments(assetAttribute: .address(wallet.address)).flatMap { (filter: $0, textEquivalent: "\(filterName)=\(wallet.address.eip55String)") }
+            case .name, .contractAddress, .symbol:
+                optionalFilter = nil
+            }
+        } else {
+            //TODO support things like "$prefix-{tokenId}"
+            optionalFilter = nil
+        }
+        guard let (filterValue, textEquivalent) = optionalFilter else { return nil }
+        guard let filterValueTypedForEventFilters = filterValue.coerceToArgumentTypeForEventFilter(parameterType) else { return nil }
+        return (filter: [filterValueTypedForEventFilters], textEquivalent: textEquivalent)
+    }
+
+    static func convertToImplicitAttribute(string: String) -> AssetImplicitAttributes? {
+        let prefix = "${"
```

Comment:
Hmm, I thought this was standard JS? Swift can't parse that more naturally?

Reply:
Not that I know of. II wanted to make it replace it verbatim, eg. "${tokenId}", but not "somePrefix#{tokenId}" for now. So it ended up looking like that without using Regex.

Any idea how to improve it?